File: oj/leetcode/242-Valid-Anagram/ValAnagram_002.java
public class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                map.put(c, map.get(c) + 1);
            } else {
                map.put(c, 1);
            }
        }
        
        for (int j = 0; j < t.length(); j++) {
            char c = t.charAt(j);
            int idx = Character.getNumericValue(c);
            if (map.containsKey(c)) {
                if (map.get(c) < 1)
                    return false;
                map.put(c, map.get(c) - 1);
            } else {
                return false;
            }
        }
        
        return true;
    }
}

File: leetcode/242-Valid-Anagram/ValAnagram_002.py
class Solution:
    # @param {string} s
    # @param {string} t
    # @return {boolean}
    def anaRepresentation(self, s):
        p = {}
        for c in s:
            if c in p:
                p[c] += 1
            else:
                p[c] = 1
        return p

    def isAnagram(self, s, t):
        if len(s) != len(t):
            return False
        p = self.anaRepresentation(s)

        for c in t:
            if c not in p:
                return False
            else:
                p[c] -= 1
                if p[c] < 0:
                    return False
        return True
